src/sentry/event_manager.py

```diff
@@ -239,32 +293,105 @@ def __init__(self, data, version='5'):
         self.version = version
 
     def normalize(self):
-        # TODO(dcramer): store http.env.REMOTE_ADDR as user.ip
-        # First we pull out our top-level (non-data attr) kwargs
         data = self.data
+        errors = data.get('errors', [])
+
+        # Before validating with a schema, attempt to cast values to their desired types
+        # so that the schema doesn't have to take every type variation into account.
+        text = six.text_type
+        fingerprint_types = six.string_types + six.integer_types + (float, )
+
+        def to_values(v):
+            return {'values': v} if v and isinstance(v, (tuple, list)) else v
+
+        casts = {
+            'environment': lambda v: text(v) if v is not None else v,
+            'fingerprint': lambda v: map(text, v) if isinstance(v, list) and all(isinstance(f, fingerprint_types) for f in v) else v,
+            'release': lambda v: text(v) if v is not None else v,
+            'dist': lambda v: text(v).strip() if v is not None else v,
+            'time_spent': lambda v: int(v) if v is not None else v,
+            'tags': lambda v: [(text(v_k.replace(' ', '-')).strip(), text(v_v).strip()) for (v_k, v_v) in dict(v).items()],
```

Comment:
this looks like the `text` is around the wrong thing? Looks like you want this to be `text(v_k).replace(...)`.

Reply:
good catch